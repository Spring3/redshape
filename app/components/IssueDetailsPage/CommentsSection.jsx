import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { remote } from 'electron';

import MarkdownEditor, { MarkdownText } from '../MarkdownEditor';
import Link from '../Link';
import DateComponent from '../Date';

const Section = styled.section`
  background: white;
  padding: 20px;
  margin-bottom: 60px;
`;

const SmallNotice = styled.p`
  font-size: 12px;
  margin-top: 0px;
  color: ${props => props.theme.minorText};

  a {
    font-size: inherit !important;
    margin-right: 5px;
  }
`;

const Comments = styled.ul`
  list-style-type: none;
  padding: 0;
  
  li {
    background: ${props => props.theme.bgDark};
    display: block;
    border-radius: 3px;
    margin-bottom: 20px;

    div.commentsHeader {
      display: flex;
      justify-content: space-between;
      padding: 20px;
      border-bottom: 2px solid ${props => props.theme.bg};

      ${({theme}) => css`
        span:first-child {
          font-weight: bold;
          color: ${theme.normalText};
        }

        span:last-child {
          font-weight: bold;
          color: ${theme.minorText};
          transition: color ease ${theme.transitionTime};
          text-align: center;

          &:hover {
            color: ${theme.normalText};
          }
        }
      `}

    }

    iframe {
      padding: 5px 20px 0px 20px;
    }
  }
`;

const CommentsForm = styled.div`
  margin-top: 40px;
  padding: 20px;
  border-radius: 3px;
  border: 2px solid ${props => props.theme.bgDark};
`;

class CommentsSection extends Component {
  sendComments = (comments) => {
    if (comments) {
      const { issueId, publishComments } = this.props;
      publishComments(issueId, comments);
    }
  }

  render() {
    const { journalEntries } = this.props;
    return (
      <Section>
        <div>
          <h2>Comments</h2>
          <Comments>
            {journalEntries.map(entry => (
              <li key={entry.id}>
                <div className="commentsHeader">
                  <span className="username">{entry.user.name}</span>
                  <DateComponent className="date" date={entry.created_on} />
                </div>
                <MarkdownText markdownText={entry.notes} />
              </li>
            ))}
          </Comments>
          <CommentsForm>
            <MarkdownEditor
              onSubmit={this.sendComments}
            />
            <p>
              <SmallNotice>
                Press&nbsp;
                {
                  remote.process.platform === 'darwin'
                    ? (<Link href="#">Cmd + Enter</Link>)
                    : (<Link href="#">Ctrl + Enter</Link>)
                }
                to send
              </SmallNotice>
            </p>
          </CommentsForm>
        </div>
      </Section>
    );
  }
}

CommentsSection.propTypes = {
  issueId: PropTypes.number.isRequired,
  journalEntries: PropTypes.arrayOf(PropTypes.shape({
    notes: PropTypes.string.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string.isRequired
    }).isRequired,
    created_on: PropTypes.string.isRequired
  }).isRequired).isRequired,
  publishComments: PropTypes.func.isRequired
};

export default CommentsSection;
