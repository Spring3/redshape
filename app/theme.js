export default {
  bg: '#FFFFFF',
  bgLight: '#F9F9F9',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#909090',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  transitionTime: '.2s'
};
