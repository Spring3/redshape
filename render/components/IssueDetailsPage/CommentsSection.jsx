import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { remote } from 'electron';

import MarkdownEditor, { MarkdownText } from '../MarkdownEditor';
import Link from '../Link';
import DateComponent from '../Date';

const Section = styled.section`
  background: white;
  padding: 20px;
  margin-bottom: 20px;
`;

const SmallNotice = styled.p`
  font-size: 12px;
  margin-top: 30px;
  color: ${props => props.theme.minorText};
  font-weight: bold;

  a {
    font-size: inherit !important;
    margin-right: 5px;
  }
`;

const Comments = styled.ul`
  list-style-type: none;
  padding: 20px 0px;
  margin: 0px;
  border-radius: 3px;
  background: ${props => props.theme.bgDark};
  
  li:first-child {
    margin-top: 20px;
  }

  li {
    margin: 40px 20px 20px 20px;
    display: flex;
    justify-content: space-around;

    div.commentsHeader {
      flex-grow: 1;
      display: flex;
      flex-direction: column;
      min-width: 20%;

      ${({theme}) => css`
        h3 {
          margin-top: 20px;
          margin-bottom: 20px;
          color: ${theme.main};
        }

        span:last-child {
          color: ${theme.minorText};
          transition: color ease ${theme.transitionTime};
          text-align: left;

          &:hover {
            color: ${theme.normalText};
          }
          margin-bottom: 20px;
        }
      `}

    }

    iframe {
      margin-left: 20px;
      padding: 5px 20px 0px 20px;
      background: ${props => props.theme.bg};
      border: 1px solid ${props => props.theme.bgDarker};
      border-radius: 3px;
      width: 74%;
      min-height: 100px;
    }
  }
`;

const CommentsForm = styled.div`
  background: ${props => props.theme.bgDark};
  padding: 20px;
  border-radius: 3px;
  border: 2px solid ${props => props.theme.bgDark};
  #commentsForm {
    background: ${props => props.theme.bg};
    padding: 20px;
    border-radius: 3px;
    border: 1px solid ${props => props.theme.bgDarker};
  }
`;

class CommentsSection extends Component {
  sendComments = (comments) => {
    if (comments) {
      const { issueId, publishComments } = this.props;
      publishComments(issueId, comments);
    }
  }

  render() {
    const { journalEntries } = this.props;
    return (
      <Section>
        <div>
          <h2>Comments</h2>
          <Comments>
            {journalEntries.map(entry => (
              <li key={entry.id}>
                <div className="commentsHeader">
                  <h3 className="username">{entry.user.name}</h3>
                  <DateComponent className="date" date={entry.created_on} />
                </div>
                <MarkdownText markdownText={entry.notes} />
              </li>
            ))}
          </Comments>
          <CommentsForm>
            <MarkdownEditor
              id="commentsForm"
              onSubmit={this.sendComments}
            />
            <div>
              <SmallNotice>
                Press&nbsp;
                {
                  remote.process.platform === 'darwin'
                    ? (<Link href="#">Cmd + Enter</Link>)
                    : (<Link href="#">Ctrl + Enter</Link>)
                }
                to send
              </SmallNotice>
            </div>
          </CommentsForm>
        </div>
      </Section>
    );
  }
}

CommentsSection.propTypes = {
  issueId: PropTypes.number.isRequired,
  journalEntries: PropTypes.arrayOf(PropTypes.shape({
    notes: PropTypes.string.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string.isRequired
    }).isRequired,
    created_on: PropTypes.string.isRequired
  }).isRequired).isRequired,
  publishComments: PropTypes.func.isRequired
};

export default CommentsSection;
