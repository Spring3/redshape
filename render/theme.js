export default {
  bg: '#FFFFFF',
  bgDark: '#F9F9F9',
  bgDarker: '#E9E9E9',
  bgDisabled: 'hsl(0,0%,95%)',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#A0A0A0',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  shadow: '#D0D0D0',
  transitionTime: '.2s'
};
