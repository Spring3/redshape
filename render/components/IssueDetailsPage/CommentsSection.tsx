import React, { useMemo } from 'react';
import { css } from '@emotion/react';
// eslint-disable-next-line
import { remote } from 'electron';
import { useTheme } from 'styled-components';

import ReactTimeAgo from 'react-time-ago';
import { MarkdownEditor, MarkdownText } from '../MarkdownEditor';
import { Link } from '../Link';
import { useOvermindActions, useOvermindState } from '../../store';
import { theme as Theme } from '../../theme';
import type { Issue } from '../../../types';
import { Flex } from '../Flex';

const styles = {
  form: (theme: typeof Theme) => css`
    width: 100%;
    background: ${theme.bgDark};
    padding: .8rem 1rem;
    border-radius: 3px;
    box-sizing: border-box;
    border: 2px solid ${theme.bgDark};
    #comments-form {
      background: ${theme.bg};
      padding: 20px;
      border-radius: 3px;
      border: 1px solid ${theme.bgDarker};
    }
  `,
  smallNoticeLink: css`
    font-size: inherit !important;
    margin-right: 5px;
  `,
  smallNotice: (theme: typeof Theme) => css`
    font-size: 12px;
    margin-top: 30px;
    color: ${theme.minorText};
    font-weight: bold;
  `,
  commentsList: (theme: typeof Theme) => css`
    width: 100%;
    padding: 20px 0px;
    margin: 0px;
    border-radius: 3px;
    background: ${theme.bgDark};
  `,
  comment: css`
    flex-grow: 1;
    margin: 1rem 1rem 1rem 1rem;
  `,
  commentHeader: css`
    flex-grow: 1;
    min-width: 20%;
    margin-right: 1rem;
  `,
  commentAuthorName: (theme: typeof Theme) => css`
    margin-top: 0px;
    margin-bottom: 10px;
    color: ${theme.main};
  `,
  commentDate: (theme: typeof Theme) => css`
    color: ${theme.minorText};
    transition: color ease ${theme.transitionTime};
    text-align: left;

    &:hover {
      color: ${theme.normalText};
    }
    margin-bottom: 20px;
  `,
  markdown: (theme: typeof Theme) => css`
    iframe {
      margin-left: 20px;
      padding: 5px 20px 0px 20px;
      background: ${theme.bg};
      border: 1px solid ${theme.bgDarker};
      border-radius: 3px;
      width: 74%;
      min-height: 100px;
    }
  `
};

type CommentsSectionProps = {
  issueId: number;
}

const CommentsSection = ({ issueId }: CommentsSectionProps) => {
  const actions = useOvermindActions();
  const state = useOvermindState();
  const theme = useTheme() as typeof Theme;

  const currentIssue = state.issues.byId[issueId];

  const getJournalEntries = (issue: Issue) => issue.journals?.filter((entry) => entry.notes) || [];

  const sendComments = async (comments?: string) => {
    if (comments) {
      actions.issues.publishComments({ issueId, comments });
    }
  };

  const journalEntries = useMemo(() => getJournalEntries(currentIssue), [currentIssue]);

  return (
    <>
      <Flex alignItems="stretch" direction="column" css={styles.commentsList(theme)}>
        {journalEntries.map((entry) => (
          <Flex justifyContent="center" css={styles.comment} key={entry.id}>
            <Flex direction="column" css={styles.commentHeader}>
              <h4 css={styles.commentAuthorName(theme)}>{entry.user.name}</h4>
              <ReactTimeAgo className="date" date={new Date(entry.createdOn)} />
            </Flex>
            <MarkdownText css={styles.markdown(theme)} name={`comment-${entry.id}`} markdownText={entry.notes as string} />
          </Flex>
        ))}
      </Flex>
      <div css={styles.form(theme)}>
        <MarkdownEditor id="comments-form" name="comments-form" onSubmit={sendComments} />
        <div>
          <p css={styles.smallNotice(theme)}>
            Press&nbsp;
            {remote.process.platform === 'darwin' ? (
              <Link css={styles.smallNoticeLink} href="#">Cmd + Enter</Link>
            ) : (
              <Link css={styles.smallNoticeLink} href="#">Ctrl + Enter</Link>
            )}
            to send
          </p>
        </div>
      </div>
    </>
  );
};

export { CommentsSection };
