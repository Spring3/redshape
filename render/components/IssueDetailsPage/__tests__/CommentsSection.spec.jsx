import React from 'react';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import CommentsSection from '../CommentsSection';
import theme from '../../../theme';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
const mockStore = configureStore([thunk]);

describe('IssueDetails => CommentsSEction componnet', () => {
  it('should match the snapshot', () => {
    const props = {
      issueId: 1,
      journalEntries: [
        {
          notes: 'Hello there',
          user: {
            id: 1,
            name: 'John Wayne'
          },
          created_on: '2011-01-01'
        },
        {
          notes: 'Does your town need a hero?',
          user: {
            id: 1,
            name: 'John Wayne'
          },
          created_on: '2011-01-02'
        }
      ],
      publishComments: jest.fn()
    };
    const state = {
      settings: {
        uiStyle: 'colors',
      },
    };
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <CommentsSection {...props} />
        </ThemeProvider>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should invoke publishComments function only if the comments are not empty', () => {
    const props = {
      issueId: 1,
      journalEntries: [
        {
          notes: 'Hello there',
          user: {
            id: 1,
            name: 'John Wayne'
          },
          created_on: '2011-01-01'
        },
        {
          notes: 'Does your town need a hero?',
          user: {
            id: 1,
            name: 'John Wayne'
          },
          created_on: '2011-01-02'
        }
      ],
      publishComments: jest.fn()
    };
    const state = {
      settings: {
        uiStyle: 'colors',
      },
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <CommentsSection {...props} />
        </ThemeProvider>
      </Provider>
    );
    const section = wrapper.find('CommentsSection').instance();
    expect(section).toBeTruthy();
    section.sendComments('');
    expect(props.publishComments).not.toHaveBeenCalled();
    section.sendComments('Comments');
    expect(props.publishComments).toHaveBeenCalledWith(props.issueId, 'Comments');
  });
});
