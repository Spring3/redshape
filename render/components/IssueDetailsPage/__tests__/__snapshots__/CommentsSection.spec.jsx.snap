// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueDetails => CommentsSection component should match the snapshot 1`] = `
.c3 {
  border-radius: 3px;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  text-align: center;
  background: #FFFFFF;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
  width: auto;
  border: 2px solid #FF7079;
  color: #FF7079;
  cursor: pointer;
}

.c3:hover,
.c3:focus {
  background: #FF7079;
  color: #FFFFFF !important;
}

.c3:hover svg,
.c3:focus svg {
  fill: #FFFFFF;
}

.c3:active {
  background: #EF6069;
}

.c7 {
  box-sizing: border-box;
  overflow: hidden;
}

.c6 {
  position: relative;
  display: inline-grid;
  grid-template-rows: 1fr;
}

.c6 span {
  float: left;
  grid-row: 1;
  grid-column: 1;
}

.c6 span:last-child {
  opacity: 0;
  display: none;
}

.c0 {
  background: white;
  padding: 20px;
  margin-bottom: 20px;
}

.c4 {
  list-style-type: none;
  padding: 20px 0px;
  margin: 0px;
  border-radius: 3px;
  background: #F9F9F9;
}

.c4 li:first-child {
  margin-top: 20px;
}

.c4 li {
  margin: 40px 20px 20px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c4 li div.commentsHeader {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 20%;
}

.c4 li div.commentsHeader h3 {
  margin-top: 20px;
  color: #FF7079;
}

.c4 li div.commentsHeader span:last-child {
  color: #A0A0A0;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
  text-align: left;
}

.c4 li div.commentsHeader span:last-child:hover {
  color: #3F3844;
}

.c4 li iframe {
  margin-left: 20px;
  padding: 5px 20px 0px 20px;
  background: #FFFFFF;
  border: 1px solid #E9E9E9;
  border-radius: 3px;
  min-height: 100px;
  width: 74%;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c1 div:last-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 div:last-child button {
  margin-right: 0.5rem;
}

.c1 div:last-child button:last-child {
  margin-right: 0;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c5 {
  padding: 8px;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  padding-left: 11px;
}

<section
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      <h2>
        Comments
      </h2>
      <div>
        <button
          className="c2 c3"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            className="mdi-icon "
            fill="currentColor"
            height={22}
            viewBox="0 0 24 24"
            width={22}
          >
            <path
              d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
            />
          </svg>
          Â Editor
        </button>
        <button
          className="c2 c3"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            className="mdi-icon "
            fill="currentColor"
            height={22}
            viewBox="0 0 24 24"
            width={22}
          >
            <path
              d="M10 11V13H18V11H10M10 17V19H14V17H10M10 5V7H22V5H10M6 7H8.5L5 3.5L1.5 7H4V20H6V7Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      className="c4"
    >
      <li
        className="c5"
      >
        <div
          className="commentsHeader"
        >
          <h3
            className="username"
          >
            John Wayne
          </h3>
          <span
            className="c6"
          >
            <span
              className="date"
            >
              Jan 01 2011
            </span>
            <span
              className="date"
            >
              Jan 01 2011
            </span>
          </span>
        </div>
        <iframe
          className="c7"
          frameBorder="0"
          onLoad={[Function]}
          srcDoc="<p>Hello there</p>"
          width="100%"
        />
      </li>
      <li
        className="c5"
      >
        <div
          className="commentsHeader"
        >
          <h3
            className="username"
          >
            John Wayne
          </h3>
          <span
            className="c6"
          >
            <span
              className="date"
            >
              Jan 02 2011
            </span>
            <span
              className="date"
            >
              Jan 02 2011
            </span>
          </span>
        </div>
        <iframe
          className="c7"
          frameBorder="0"
          onLoad={[Function]}
          srcDoc="<p>Does your town need a hero?</p>"
          width="100%"
        />
      </li>
    </ul>
  </div>
</section>
`;
